ember-server/src/main/scala/org/http4s/ember/server/internal/WebSocketHelpers.scala

```diff
@@ -0,0 +1,213 @@
+    }
+
+    wsResponse
+      .flatMap { res =>
+        ServerHelpers.send(socket)(Some(req), res, idleTimeout, onWriteFailure).void
+      } >> runConnection(socket, ctx, buffer, receiveBufferSize, idleTimeout)
+  }
+
+  private def runConnection[F[_]](
+      socket: Socket[F],
+      ctx: WebSocketContext[F],
+      buffer: Array[Byte],
+      receiveBufferSize: Int,
+      idleTimeout: Duration)(implicit F: Concurrent[F]): F[Unit] = {
+    val read: Read[F] = socket.read(receiveBufferSize, durationToFinite(idleTimeout))
+    val frameTranscoder = new FrameTranscoder(false)
+
+    // TODO: make sure error semantics are correct and that resources are properly cleaned up
+    // TODO: consider write/read failures and effect on outer connection
+    // TODO: there is some shared code here with ServerHelpers
+    val writer = ctx.webSocket.send
+      .flatMap { frame =>
+        // TODO: frameToBuffer can throw
+        Stream
+          .iterable(frameTranscoder.frameToBuffer(frame).map { buffer =>
+            // TODO: improve
+            val bytes = new Array[Byte](buffer.remaining())
+            buffer.get(bytes)
+            Chunk.bytes(bytes)
+          })
+          .flatMap(Stream.chunk(_))
+      }
+      .through(socket.writes(durationToFinite(idleTimeout)))
+
+    val reader = (Stream.chunk(Chunk.bytes(buffer)) ++ readStream(read))
+      .through(decodeFrames(frameTranscoder))
+      .through(ctx.webSocket.receive)
+
+    reader
+      .concurrently(writer)
+      .drain
+      .compile
+      .drain
+      .attempt
+      .flatMap {
+        case Left(err) =>
+          err.printStackTrace()
+          F.unit
+        case Right(()) =>
+          println("Connection ended")
+          F.unit
+      }
+  }
+
+  private def decodeFrames[F[_]](frameTranscoder: FrameTranscoder)(implicit
+      F: Concurrent[F]): Pipe[F, Byte, WebSocketFrame] = stream => {
+    def go(rest: Stream[F, Byte], acc: Array[Byte]): Pull[F, WebSocketFrame, Unit] =
+      rest.pull.uncons.flatMap {
+        case Some((chunk, next)) =>
+          val buffer = acc ++ chunk.toArray
+          val byteBuffer = ByteBuffer.wrap(buffer)
+          Pull
+            .attemptEval(F.delay(frameTranscoder.bufferToFrame(byteBuffer)))
+            .flatMap {
+              case Right(value) =>
+                // TODO: value is nullable
+                val remaining = new Array[Byte](byteBuffer.remaining())
+                byteBuffer.get(remaining)
+                Pull.output1(value) >> go(next, remaining)
+              case Left(err) =>
+                // TODO: figure out what to do here
+                println(err)
+                Pull.done
+            }
+        case None =>
+          // TODO: figure out what to do here
+          println("done")
+          Pull.done
+      }
+
+    go(stream, Array.emptyByteArray).void.stream
+  }
+
+  private def clientHandshake[F[_]](req: Request[F]): Either[ClientHandshakeError, String] = {
```

Comment:
It made more sense to re-implement this more cleanly rather than rely on the existing handshaker. We can push it to the other package perhaps? I feel like other backends with the exception of Blaze wouldn't really use it

Reply:
I think servlets could.  The JSR web socket does not support our model, because we can't conditionally upgrade HTTP.  But the upgrade handler, now implemented in Jetty 10, gives us low-level control that might work just about like this!